src/QSymbolicsBase/rules.jl

```diff
@@ -40,7 +47,83 @@ end
 
 FLATTEN_RULES = [
     @rule(~x::isnotflat_precheck(⊗) => flatten_term(⊗, ~x)),
-    @rule ⊗(~~xs::hasscalings) => prefactorscalings_rule(xs)
+    @rule ⊗(~~xs::hasscalings) => prefactorscalings_rule(xs) # Used to perform (a*|k⟩) ⊗ (b*|l⟩) → (a*b) * (|k⟩⊗|l⟩) 
 ]
 
 tensor_simplify = Fixpoint(Chain(FLATTEN_RULES))
+
+"""Quantum circuit identities"""
+_isXX(x) = x isa SApplyOp && x.terms == [X,X] # work on this with identity op
+_isYY(x) = x isa SApplyOp && x.terms == [Y,Y]
+_isZZ(x) = x isa SApplyOp && x.terms == [Z,Z]
+_isXY(x) = x isa SApplyOp && x.terms == [X,Y]
+_isYZ(x) = x isa SApplyOp && x.terms == [Y,Z]
+_isZX(x) = x isa SApplyOp && x.terms == [Z,X]
+_isYX(x) = x isa SApplyOp && x.terms == [Y,X]
+_isZY(x) = x isa SApplyOp && x.terms == [Z,Y]
+_isXZ(x) = x isa SApplyOp && x.terms == [X,Z]
+_isHXH(x) = x isa SApplyOp && x.terms == [H,X,H]
+_isHYH(x) = x isa SApplyOp && x.terms == [H,Y,H]
+_isHZH(x) = x isa SApplyOp && x.terms == [H,Z,H]
+
+
+CIRCUIT_RULES = [
+    @rule(~x::_isXX => I),
```

Comment:
I would be very surprised if there isn't a simpler cleaner way to implement these. Creating a predicate to be used in a single rule seems much too cumbersome. This is exactly what pattern matching is supposed to simplify.

Reply:
well, I checked, and I am indeed very surprised how convoluted it is, but it can still be simpler (and more "pattern match"-ie)  than this version.

E.g. consider this (misusing `tensor` instead of multiplication because the current version of QSymbolics does not have an operation multiplication implemented)

```
julia> _isa(T) = x->isa(x,T)
_isa (generic function with 1 method)

julia> r = @rule ~O₁::_isa(XGate) ⊗ ~O₂::_isa(ZGate) => -im*Y
~(O₁::_isa(XGate)) ⊗ ~(O₂::_isa(ZGate)) => -im * Y

julia> r(X ⊗ Z)
0 - 1imY
```

Reply:
That code doesn't work for me: what you wrote is very similar to the example I showed you on Tuesday. When I define that rule, `r(X⊗Z)` gives `nothing`.

Reply:
weird... on this exact branch I get the following (copying the status of installed projects for completeness)

```
(quantumjulia) pkg> st
Status `~/Documents/ScratchSpace/quantumjulia/Project.toml`
⌅ [c3fe647b] AbstractAlgebra v0.40.9
  [cbdf2221] AlgebraOfGraphics v0.6.19
  [2c1a90cd] BPGates v1.1.0 `BPGates.jl`
  [6e4b80f9] BenchmarkTools v1.5.0
  [824d6782] Bonito v3.1.1
  [13f3f980] CairoMakie v0.12.2
  [9961bab8] Cbc v1.2.0
  [861a8166] Combinatorics v1.0.2
  [6ed1e86c] ConcurrentSim v1.4.1 `ConcurrentSim.jl`
  [992eb4ea] CondaPkg v0.2.22
  [a93c6f00] DataFrames v1.6.1
  [864edb3b] DataStructures v0.18.20
  [459566f4] DiffEqCallbacks v3.6.2
  [31c24e10] Distributions v0.25.109
  [ffbed154] DocStringExtensions v0.9.3
  [e30172f5] Documenter v1.4.1
  [5789e2e9] FileIO v1.16.3
  [e9467ef8] GLMakie v0.10.2
  [c27321d9] Glob v1.3.1
  [1ecd5474] GraphMakie v0.5.11
  [86223c79] Graphs v1.11.1
  [06909019] GraphsFlows v0.1.1
  [c3af3a8c] GraphsMatching v0.2.0
  [cd3eb016] HTTP v1.10.8
  [7073ff75] IJulia v1.24.2
  [c3a54625] JET v0.9.3
  [0f8b85d8] JSON3 v1.14.0
  [4076af6c] JuMP v1.22.1
  [98e50ef6] JuliaFormatter v1.0.56
  [3c486d74] LDPCDecoders v0.3.1 `LDPCDecoders.jl`
  [b964fa9f] LaTeXStrings v1.3.1
  [1914dd2f] MacroTools v0.5.13
  [ee78f7c6] Makie v0.21.2
⌅ [2edaba10] Nemo v0.43.3
  [46757867] NetworkLayout v0.4.6
  [8913a72c] NonlinearSolve v3.12.4
  [f1435218] Oscar v1.0.3
  [df9a0d86] Oxygen v1.5.10
  [e4faabce] PProf v3.1.0
  [2ae35dd2] Permutations v0.4.21
  [32113eaa] PkgBenchmark v0.2.12
  [91a5bcdd] Plots v1.40.4
  [21216c6a] Preferences v1.4.3
  [92933f4c] ProgressMeter v1.10.0
  [438e738f] PyCall v1.96.4
  [17f5de1a] PyQDecoders v0.2.0 `PyQDecoders.jl`
  [6099a3de] PythonCall v0.9.20
  [b0d11df0] Quantikz v1.3.1 `Quantikz.jl`
  [0525e862] QuantumClifford v0.9.4 `QuantumClifford.jl`
  [5717a53b] QuantumInterface v0.3.4 `QuantumInterface.jl`
  [6e0679c1] QuantumOptics v1.0.16 `QuantumOptics.jl`
  [4f57444f] QuantumOpticsBase v0.4.22 `QuantumOpticsBase.jl`
  [efa7fd63] QuantumSymbolics v0.3.0 `QuantumSymbolics.jl`
  [c5292f4c] ResumableFunctions v0.6.10 `ResumableFunctions.jl`
  [295af30f] Revise v3.5.14
  [276daf66] SpecialFunctions v2.4.0
  [860ef19b] StableRNGs v1.0.2
  [856f2bd8] StructTypes v1.10.0
  [8e1ec7a9] SumTypes v0.5.8 `~/.julia/dev/SumTypes`
  [1b6eb727] SwaggerMarkdown v0.2.1
  [24249f21] SymPy v2.1.1
  [0c5d862f] Symbolics v5.30.3
  [276b4fcb] WGLMakie v0.10.2
  [ddb6d928] YAML v0.4.11
  [37e2e46d] LinearAlgebra
  [56ddb016] Logging
  [8dfed614] Test
Info Packages marked with ⌅ have new versions available but compatibility constraints restrict them from upgrading. To see why use `status --outdated`

julia> using QuantumSymbolics, Symbolics

julia> _isa(T) = x->isa(x,T)
_isa (generic function with 1 method)

julia> r = @rule ~O₁::_isa(XGate) ⊗ ~O₂::_isa(ZGate) => -im*Y
~(O₁::_isa(XGate)) ⊗ ~(O₂::_isa(ZGate)) => -im * Y

julia> r(X ⊗ Z)
0 - 1imY

julia> versioninfo()
Julia Version 1.10.4
Commit 48d4fd48430 (2024-06-04 10:41 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 32 × AMD Ryzen 9 7950X 16-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, znver3)
Threads: 32 default, 1 interactive, 16 GC (on 32 virtual cores)
```

Reply:
Ah, I wasn't using the latest version of `Symbolics`. Works now, thanks!

Reply:
oh, if that was a source of things not working, we should appropriately update the compat in Project.tom.